packages/router/src/utils/collection.ts

```diff
@@ -20,12 +20,14 @@ export function shallowEqualArrays(a: any[], b: any[]): boolean {
 }
 
 export function shallowEqual(a: Params, b: Params): boolean {
-  // Casting Object.keys return values to include `undefined` as there are some cases
-  // in IE 11 where this can happen. Cannot provide a test because the behavior only
-  // exists in certain circumstances in IE 11, therefore doing this cast ensures the
-  // logic is correct for when this edge case is hit.
-  const k1 = Object.keys(a) as string[] | undefined;
-  const k2 = Object.keys(b) as string[] | undefined;
+  // Previously cast Object.keys return values to include `undefined` as there
+  // are some cases in IE 11 where this can happen. Cannot provide a test
+  // because the behavior only exists in certain circumstances in IE 11,
+  // therefore doing this cast ensures the logic is correct for when this edge
+  // case is hit. Adding a check for a and b being defined because in Edge
+  // sometimes Object.keys throws an error.
+  const k1 = a ? Object.keys(a) : undefined;
+  const k2 = b ? Object.keys(b) : undefined;
   if (!k1 || !k2 || k1.length != k2.length) {
     return false;
```

Comment:
Could we just bail out early if a or b is `undefined`?

```suggestion
  if (!a || !b) {
    return false;
  }
  const k1 = Object.keys(a);
  const k2 = Object.keys(b);
  if (k1.length != k2.length) {
    return false;
```

Or is it still possible for `Object.keys(x)` to return `undefined` even if `x` is not undefined?

Reply:
I think the old comment is indicating that `Object.keys(x)` would sometimes return `undefined` even if `x` is not.

Reply:
Since this change simplifies the casting of k1 and k2 and has been tested already I'd prefer to keep it as it. Might be able to simplify this in the future once folks are off older versions of Edge